File: src/main/java/com/DavidM1A2/afraidofthedark/common/spell/component/effect/SpellEffectDig.java
package com.DavidM1A2.afraidofthedark.common.spell.component.effect;

import com.DavidM1A2.afraidofthedark.common.constants.ModSpellEffects;
import com.DavidM1A2.afraidofthedark.common.spell.Spell;
import com.DavidM1A2.afraidofthedark.common.spell.component.effect.base.AOTDSpellEffect;
import com.DavidM1A2.afraidofthedark.common.spell.component.effect.base.SpellEffectEntry;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.Entity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

/**
 * Dig effect digs a block
 */
public class SpellEffectDig extends AOTDSpellEffect
{
    /**
     * Constructor just calls super, there's nothing else to init
     */
    public SpellEffectDig()
    {
        super();
    }

    /**
     * Gets the cost of the effect given the state of the effect
     *
     * @return The cost of dig is 14
     */
    @Override
    public double getCost()
    {
        return 14;
    }

    /**
     * Performs the dig effect against a given entity which removes the block under them
     *
     * @param spell The spell that caused the effect
     * @param spellStageIndex The spell stage that this effect is a part of
     * @param effectIndex The effect slot that this effect is in
     * @param entityHit The entity that the effect should be applied to
     */
    @Override
    public void performEffect(Spell spell, int spellStageIndex, int effectIndex, Entity entityHit)
    {
        // Digs the block under the player
        BlockPos blockPos = entityHit.getPosition().down();
        if (this.canBlockBeDestroyed(entityHit.world, blockPos))
        {
            entityHit.world.destroyBlock(blockPos, true);
        }
    }

    /**
     * Performs the effect against a given block in the world
     *
     * @param spell The spell that caused the effect
     * @param spellStageIndex The spell stage that this effect is a part of
     * @param effectIndex The effect slot that this effect is in
     * @param world The world the effect is being fired in
     * @param position The position the effect is being performed at
     */
    @Override
    public void performEffect(Spell spell, int spellStageIndex, int effectIndex, World world, BlockPos position)
    {
        // Digs the block at the position
        if (this.canBlockBeDestroyed(world, position))
        {
            world.destroyBlock(position, true);
        }
    }

    /**
     * Tests if a given block can be broken with a dig spell
     *
     * @param world The world the block is in
     * @param blockPos The pos the block is at
     * @return True if the block can be destroyed, false otherwise
     */
    private boolean canBlockBeDestroyed(World world, BlockPos blockPos)
    {
        IBlockState blockState = world.getBlockState(blockPos);
        return blockState.getBlock().getBlockHardness(blockState, world, blockPos) != -1;
    }

    /**
     * Should get the SpellEffectEntry registry's type
     *
     * @return The registry entry that this component was built with, used for deserialization
     */
    @Override
    public SpellEffectEntry getEntryRegistryType()
    {
        return ModSpellEffects.DIG;
    }
}

File: src/main/java/com/davidm1a2/afraidofthedark/common/spell/component/effect/SpellEffectDig.kt
package com.davidm1a2.afraidofthedark.common.spell.component.effect

import com.davidm1a2.afraidofthedark.common.constants.Constants
import com.davidm1a2.afraidofthedark.common.spell.component.DeliveryTransitionState
import com.davidm1a2.afraidofthedark.common.spell.component.SpellComponentInstance
import com.davidm1a2.afraidofthedark.common.spell.component.effect.base.AOTDSpellEffect
import com.davidm1a2.afraidofthedark.common.spell.component.effect.base.SpellEffect
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

/**
 * Dig effect digs a block
 */
class SpellEffectDig : AOTDSpellEffect(ResourceLocation(Constants.MOD_ID, "dig"))
{
    /**
     * Performs the effect
     *
     * @param state The state that the spell is in
     * @param instance The instance of the effect
     */
    override fun procEffect(state: DeliveryTransitionState, instance: SpellComponentInstance<SpellEffect>)
    {
        val world = state.world
        val entity = state.getEntity()
        if (entity != null)
        {
            // Digs the block under the entity
            val blockPos = entity.position.down()
            if (canBlockBeDestroyed(world, blockPos))
            {
                createParticlesAt(1, 3, entity.positionVector, entity.dimension)
                world.destroyBlock(blockPos, true)
            }
        }
        else
        {
            // Digs the block at the position
            val position = state.blockPosition
            if (canBlockBeDestroyed(world, position))
            {
                createParticlesAt(1, 3, state.position, world.provider.dimension)
                world.destroyBlock(position, true)
            }
        }
    }

    /**
     * Tests if a given block can be broken with a dig spell
     *
     * @param world    The world the block is in
     * @param blockPos The pos the block is at
     * @return True if the block can be destroyed, false otherwise
     */
    private fun canBlockBeDestroyed(world: World, blockPos: BlockPos): Boolean
    {
        val blockState = world.getBlockState(blockPos)
        @Suppress("DEPRECATION")
        return blockState.block.getBlockHardness(blockState, world, blockPos) != -1f
    }

    /**
     * Gets the cost of the effect given the state of the effect
     *
     * @param instance The instance of the spell effect to grab the cost of
     * @return The cost of dig is 14
     */
    override fun getCost(instance: SpellComponentInstance<SpellEffect>): Double
    {
        return 14.0
    }
}